### PR TITLE
Added the story mode with 2 methods and add min comment lenght

### PR DESCRIPTION
**Pull request by electro199**
Added the story mode with two options! The first option is only a screenshot of the post description (looks ugly with long posts). The second option generates images for every sentence.  

I also added minimum comment length sorting feature

# Checklist:

- [x] I am pushing changes to the **develop** branch
- [x] I am using the recommended development environment
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have formatted and linted my code using python-black and pylint
- [x] I have cleaned up unnecessary files
- [x] My changes follow the existing code-style
- [x] My changes are relevant to the project
- [x] My changes generate no new warnings

# Any other information (e.g how to test the changes)

It's a big feature and it needs to be tested a lot.

### CODE REVIEW

**Comment by electro199** [comment]
I have solve conflicts

File: video_creation/screenshot_downloader.py [hunk 1]
```diff
@@ -6,15 +6,10 @@
 import translators as ts
 from playwright.sync_api import sync_playwright, ViewportSize
 from rich.progress import track
-
-from utils import settings
```
Comment (OpenSourceSimon) [inline video_creation/screenshot_downloader.py:10]: Why did you remove this line?

File: utils/subreddit.py [hunk 2]
```diff
@@ -39,6 +39,9 @@ def get_subreddit_undone(submissions: list, subreddit, times_checked=0):
                 f'This post has under the specified minimum of comments ({settings.config["reddit"]["thread"]["min_comments"]}). Skipping...'
             )
             continue
+        if  settings.config['settings']['storymode'] :
+                if  not submission.is_self   :
+                    continue
```
Comment (OpenSourceSimon) [inline utils/subreddit.py:44]: Where is this if statement for and why is it only for story mode?

File: TTS/engine_wrapper.py [hunk 3]
```diff
@@ -18,8 +18,7 @@
 from utils.console import print_step, print_substep
 from utils.voice import sanitize_text
 
-DEFAULT_MAX_LENGTH: int = 50  # video length variable
-
+DEFAULT_MAX_LENGTH: int = 40 # video length variable
```
Comment (OpenSourceSimon) [inline TTS/engine_wrapper.py:21]: Can you change this back to 50?

**Review by OpenSourceSimon — COMMENTED** [review]
Thanks for the PR, but I have a couple of questions. Also, this needs to be tested

File: TTS/engine_wrapper.py (hunk 3 above)
Comment (electro199) [reply TTS/engine_wrapper.py:21]: Yesv

File: utils/subreddit.py (hunk 2 above)
Comment (electro199) [reply utils/subreddit.py:44]: It will check if story mode is enabled and check for the post for is_self if not it will check next post. It can also called strict story mode i had some version with non strict story mode codes

**Review by OpenSourceSimon — APPROVED** [review]

File: video_creation/final_video.py [hunk 4]
```diff
@@ -75,10 +75,20 @@ def make_final_video(
...
+    
     # Gather all audio clips
-    audio_clips = [AudioFileClip(f"assets/temp/{id}/mp3/{i}.mp3") for i in range(number_of_clips)]
-    audio_clips.insert(0, AudioFileClip(f"assets/temp/{id}/mp3/title.mp3"))
+    if settings.config["settings"]["storymode"]:
+        if settings.config["settings"]["storymodemethod"] == 0:
+            audio_clips = [AudioFileClip(f"assets/temp/{id}/mp3/title.mp3")]
+            audio_clips.insert(1,AudioFileClip(f"assets/temp/{id}/mp3/postaudio.mp3"))
+        elif settings.config["settings"]["storymodemethod"] == 1:
+            #here work is not done14
```
Comment (OpenSourceSimon) [inline video_creation/final_video.py:85]: Can you fix this?

**Review by OpenSourceSimon — CHANGES_REQUESTED** [review]

File: utils/imagenarator.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        y += line_height + padding
+
+
+def imagemaker( theme,reddit_obj,
+                txtclr,
+                padding=5
+                ):
+    texts=reddit_obj['thread_post']
+    id = re.sub(r"[^\w\s-]", "", reddit_obj["thread_id"])
+    font=ImageFont.truetype("C:\\fonts\\NotoSans-Regular.ttf", 20)
```
Comment (OpenSourceSimon) [inline utils/imagenarator.py:26]: Why is this an absolute path that only works on the C drive? Maybe get the path: https://stackoverflow.com/questions/51520/how-to-get-an-absolute-file-path-in-python

**Review by OpenSourceSimon — CHANGES_REQUESTED** [review]

**Comment by JasonLovesDoggo** [comment]
Try using https://moviepy-tburrows13.readthedocs.io/en/latest/_modules/moviepy/video/fx/scroll.html

Instead of multiple screenshots

**Comment by electro199** [comment]
> Try using https://moviepy-tburrows13.readthedocs.io/en/latest/_modules/moviepy/video/fx/scroll.html
> 
> Instead of multiple screenshots

yes possible it looks more like comment mode

**Comment by Shevly** [comment]
When will this be released? Or How can I at least get the beta version of this to be working?

**Comment by OpenSourceSimon** [comment]
> When will this be released? Or How can I at least get the beta version of this to be working?

Hey, I hope we can release this soon. But it needs to be tested a lot because it's a big feature. I would recommend to join Lewis' Discord server, more information about testing the story mode is there!

File: utils/version.py [hunk 6]
```diff
@@ -3,13 +3,17 @@
...
 
 
-def checkversion(__VERSION__):
+def checkversion(__VERSION__:str):
     response = requests.get("https://api.github.com/repos/elebumm/RedditVideoMakerBot/releases/latest")
     latestversion = response.json()["tag_name"]
     if __VERSION__ == latestversion:
         print_step(f"You are using the newest version ({__VERSION__}) of the bot")
         return True
+    #FOR ERROR HANDLING, SO VERSION CAN EASILY BE DETECTED
```
Comment (OpenSourceSimon) [inline utils/version.py:12]: Now there isn't a notification that you're using an older version. I do like the idea of showing the beta version message. Maybe do it with a higher or lower version if statement?

File: requirements.txt [hunk 7]
```diff
@@ -14,3 +14,5 @@ pyttsx3==2.90
 Pillow~=9.1.1
 tomlkit==0.11.4
 Flask==2.2.2
+spacy==3.4.1
+https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.0.0/en_core_web_sm-3.0.0.tar.gz
```
Comment (OpenSourceSimon) [inline requirements.txt:18]: Can you remove this URL?

File: video_creation/final_video.py [hunk 8]
```diff
@@ -75,10 +75,21 @@ def make_final_video(
...
```
Comment (OpenSourceSimon) [inline video_creation/final_video.py:85]: Can you work on this so it's done?

File: video_creation/final_video.py [hunk 9]
```diff
@@ -97,29 +108,44 @@ def make_final_video(
...
-    #        ImageClip("assets/png/title.png")
-    #        .set_duration(audio_clips[0].duration + audio_clips[1].duration)
-    #        .set_position("center")
-    #        .resize(width=W - 100)
-    #        .set_opacity(float(opacity)),
-    #    )
-    # else: story mode stuff
+    if settings.config["settings"]["storymode"]:
+        if settings.config["settings"]["storymodemethod"] == 0:
+            if os.path.exists(f"assets/temp/{id}/png/story_content.png"):# else: story mode stuff
```
Comment (OpenSourceSimon) [inline video_creation/final_video.py:113]: The comment here can be removed I think?

File: video_creation/screenshot_downloader.py [hunk 10]
```diff
@@ -26,81 +30,111 @@ def download_screenshots_of_reddit_posts(reddit_object: dict, screenshot_num: in
...
+        txtcolor=(240,240,240)
+    else:
+        cookie_file = open("./video_creation/data/cookie-light-mode.json", encoding="utf-8")
+        bgcolor=(255,255,255,255)
+        txtcolor=(0,0,0)
+    if settings.config["settings"]["storymode"] :
+                # if settings.config["settings"]["storymodemethode"] == 0:
+                #     story=True
+                        
+                if settings.config["settings"]["storymodemethod"] :
```
Comment (OpenSourceSimon) [inline video_creation/screenshot_downloader.py:135]: This needs to be if == 1, right?

**Review by OpenSourceSimon — CHANGES_REQUESTED** [review]
Awesome PR, can't test it right now. Could you look at the things that I commented on? When that's fixed, I'll be happy to test it.

**Comment by JasonLovesDoggo** [comment]
I definitely agree with Simon here

**Comment by electro199** [comment]
> Awesome PR, can't test it right now. Could you look at the things that I commented on? When that's fixed, I'll be happy to test it.

ok i will consider it

File: main.py [hunk 11]
```diff
@@ -102,5 +103,8 @@ def shutdown():
         print_markdown("Please check your credentials in the config.toml file")
 
         shutdown()
-
+    except Exception as err :
+        
+        print_step('looks like some thing gone wrong in testing builts it happens report on us discord')
```
Comment (OpenSourceSimon) [inline main.py:108]:
```suggestion
        print_step('Sorry, something went wrong with this test version! Try again, and feel free to report this issue at GitHub or the Discord community.')
```

File: requirements.txt (hunk 7 above)
Comment (OpenSourceSimon) [inline requirements.txt:18]:
```suggestion
```

File: video_creation/screenshot_downloader.py (hunk 10 above)
Comment (OpenSourceSimon) [inline video_creation/screenshot_downloader.py:135]:
```suggestion
                if settings.config["settings"]["storymodemethod"] == 1:
```

File: reddit/subreddit.py [hunk 12]
```diff
@@ -40,8 +41,9 @@ def get_subreddit_threads(POST_ID: str):
             check_for_async=False,
         )
     except ResponseException as e:
-        if e.response.status_code == 401:
-            print("Invalid credentials - please check them in config.toml")
+        match e.response.status_code:
+            case 401:
+                print("Invalid credentials - please check them in config.toml")
```
Comment (OpenSourceSimon) [inline reddit/subreddit.py:46]: Can you revert this change? We removed the match statement because it's only supported by >Python 3.10

**Review by OpenSourceSimon — CHANGES_REQUESTED** [review]
Looks good except this one thing. I'll test as soon as possible.

**Comment by OpenSourceSimon** [comment]
I'll commit some code which mostly includes reformatting from Black and a couple of bug fixes. There definitely needs to be a check for posts that don't support story mode, when story mode is enabled.

**Comment by gabx03** [comment]
first it says this: 

'ffmpeg' is not recognized as an internal or external command

then this: 

\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\moviepy\video\io\ffmpeg_reader.py", line 270, in ffmpeg_parse_infos
    raise IOError(("MoviePy error: the file %s could not be found!\n"
OSError: MoviePy error: the file assets/temp/z0x8ye/mp3/title.mp3 could not be found!
Please check that you entered the correct path.


if you need more traceback let me know.

**Comment by gabx03** [comment]
I managed to fix that and now i get 'File not found: assets/temp/id/mp3/part.mp3' while using TikTok TTS, because no mp3 files were being created. Probably because TikTok TTS is deprecated at the current version. Then 'waiting for selector "#t1_ixbkbc2' while using pytts, because mp3 files were being created at least. @electro199

**Comment by OpenSourceSimon** [comment]
> I managed to fix that and now i get 'File not found: assets/temp/id/mp3/part.mp3' while using TikTok TTS, because no mp3 files were being created. Probably because TikTok TTS is deprecated at the current version. Then 'waiting for selector "#t1_ixbkbc2' while using pytts, because mp3 files were being created at least. @electro199

The TikTok TTS currently doesn't work. Try streamlabspolly or Google Translate

**Review by OpenSourceSimon — APPROVED** [review]
I've tested it and it works perfect! If someone else reviews this too, we can squash merge it!

File: TTS/engine_wrapper.py [hunk 13]
```diff
@@ -18,7 +18,7 @@
 from utils.console import print_step, print_substep
 from utils.voice import sanitize_text
 
-DEFAULT_MAX_LENGTH: int = 50  # video length variable
+DEFAULT_MAX_LENGTH: int = 29  # video length variable
```
Comment (OpenSourceSimon) [inline TTS/engine_wrapper.py:21]: Needs to be set back when merging

File: main.py [hunk 14]
```diff
@@ -19,10 +20,10 @@
     get_background_config,
 )
 from video_creation.final_video import make_final_video
-from video_creation.screenshot_downloader import download_screenshots_of_reddit_posts
+from video_creation.screenshot_downloader import get_screenshots_of_reddit_posts
 from video_creation.voices import save_text_to_mp3
 
-__VERSION__ = "2.4.2"
+__VERSION__ = "2.5.4"
```
Comment (OpenSourceSimon) [inline main.py:26]:
```suggestion
__VERSION__ = "2.5.0"
```

File: TTS/engine_wrapper.py (hunk 13 above)
Comment (OpenSourceSimon) [inline TTS/engine_wrapper.py:21]:
```suggestion
DEFAULT_MAX_LENGTH: int = 50 # video length variable
```

**Review by OpenSourceSimon — COMMENTED** [review]
Thanks for your PR! Can you fix the things I reviewed? Then I'll finally merge

**Review by OpenSourceSimon — APPROVED** [review]
